### PR TITLE
Require root packages when defining nested packages in wit

### PR DESCRIPTION
**Pull request by macovedj**
This PR is an initial implementation of the wit functionality described [here](https://github.com/WebAssembly/component-model/pull/380)

Now, whenever defining packages as `package {...}`, it will be required that some file has `package namespace:name;` as a header somewhere in the project.  Thus there is a single root package, and the other package declarations are nested.  It also supports intermixing interface and world definitions between nested package definitions across files, and the packages in a deps folder can also nest packages, which are usable by the root package definition and its nested packages

### CODE REVIEW

File: tests/cli/wit-directory-output-valid.wit
Comment (alexcrichton) [inline tests/cli/wit-directory-output-valid.wit:3]: Mind transforming this test to be valid with a `package foo:bar;` at the top?

File: tests/cli/multiple-packages-multiple-worlds.wit.stderr
Comment (alexcrichton) [inline tests/cli/multiple-packages-multiple-worlds.wit.stderr:1]: Could this error be made a bit more descriptive in that it's just the main package which contains no worlds?

File: tests/cli/multiple-packages-one-world.wit
Comment (alexcrichton) [inline tests/cli/multiple-packages-one-world.wit:3]: This test feels a bit off, now that there's the concept of a "main package" what this test was previously testing isn't quite as applicable any more and it should fail saying the main package has no worlds in it. The test could also be updated to pass a `--world` argument perhaps.

File: tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit
Comment (alexcrichton) [inline tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit:1]: Could this test be preserved with a new package header at the top?

File: crates/wit-component/src/encoding.rs
Comment (alexcrichton) [inline crates/wit-component/src/encoding.rs:2158]: You might already be planning to get to this, but `select_world` can be updated to taking a `PackageId` now instead of a list which will have a ripple effect throughout this repository.

**Review by alexcrichton — COMMENTED** [review]
Some initial thoughts jotted down, but I'll hold off on a full review until this is ready.

File: tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit
Comment (macovedj) [reply tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit:1]: Yeah I've been trying to decide what to do about this test, as well as what the desired behavior even is for this command.

Right now `DecodedWasm` doesn't differentiate between whether packages are from a deps folder or nested.  We could add this differentiation.
 
As I see it at the moment, there are a few options available:

1.) Capture if a package is nested or a deps folder package when we resolve, so we know how the package should be printed.  I think this would also require storing this info in binaries?  Which feels like something we may not want to do?

2.) Every package that isn't the root is nested... no deps folder is printed.

3.) Every package that isn't the root is printed in a deps folder... no nested packages (though packages in the original deps folder can have nested package definitions, so those would be nested still)

4.) Flags determine behavior

Maybe I'm missing an obvious alternative...

File: tests/cli/wit-directory-output-valid.wit
Comment (macovedj) [reply tests/cli/wit-directory-output-valid.wit:3]: This test has the same uncertainties as the one below testing the `--out-dir` flag.

File: tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit
Comment (alexcrichton) [reply tests/cli/wit-directory-output-in-deps-folder-with-multiple-version.wit:1]: Ah good points, but I think it's ok to create a `deps` folder when there wasn't one before. Organization of WIT might vary but the tooling here splatting out a single way to represent things I think makes sense. This is mostly just a test that whatever `--out-dir` produces is readable by `component wit`.

Implementation-wise I think it's fine to either shove everything into a `deps/*.wit` file or even just use `package { ... }` for all deps in the root folder. The whole `--out-dir` argument dosn't make a ton of sense any more nowadays that there's multi-package-forms-in-one-file but I figure  it's still good to not break it in the meantime.

File: crates/wasm-wave/src/value/wit.rs
Comment (alexcrichton) [inline crates/wasm-wave/src/value/wit.rs:158]: Mind updating these tests to using `resolve.push_str` instead of using `UnresolvedPackageGroup`?

File: crates/wit-component/src/printing.rs
Comment (alexcrichton) [inline crates/wit-component/src/printing.rs:54]: Could this change to taking a single `PackageId` instead of a list? I think it's a bit confusing to have the "main" one be at the end and it would help emphasize that this only prints a single package at a time. 

I'm imagining that internally it would use `LiveTypes` to calculate the set of packages that this package depends on and then all of those packages would be printed in curly braces at the end. That way the semantic meaning of the output is the same but this is a way to print a whole package as a self-contained entity.

As an optional part the "print all deps too" logic could be controlled by a `bool` parameter, but that's ok to defer to later.

File: crates/wit-parser/src/resolve.rs
Comment (alexcrichton) [inline crates/wit-parser/src/resolve.rs:1493]: How come this changed from `UnresolvedPackage` to `&mut`? The previous signature helped reflect that ownership was transferred and I think it'd be best to keep it if possible.

File: crates/wit-parser/src/resolve.rs
Comment (alexcrichton) [inline crates/wit-parser/src/resolve.rs:1512]: As I read over this I don't think that this is quite right, but not because of this PR. This is actually a preexisting issue today. Basically this isn't quite right because below the packages need to be merged rather than appended to otherwise duplicates might show up.

That being said I'll try to fix this in a follow up since it's a preexisting issue independent of this PR.

File: crates/wit-parser/src/resolve.rs
Comment (alexcrichton) [inline crates/wit-parser/src/resolve.rs:366]: Could this preserve the `Vec` return value? This looks like it's returning an error on duplicate packages but most of the rest of WIT processing ends up merging things together so that may be best to do here instead.

File: crates/wit-parser/src/resolve.rs
Comment (alexcrichton) [inline crates/wit-parser/src/resolve.rs:332]: Is it possibly to rely on `sort_unresolved_packages` here instead of duplicating some parts of it? Ideally there'd be just one "sort this soup" method but I also realize that there's a lot of entry points here and may make that complicated to write.

File: crates/wit-parser/src/lib.rs
Comment (alexcrichton) [inline crates/wit-parser/src/lib.rs:129]: Mind adding documentation to these two fields? (adding to `main` and updating the ones on `nested`)

File: crates/wit-parser/src/ast.rs
Comment (alexcrichton) [inline crates/wit-parser/src/ast.rs:102]: Could this error get deferred to somewhere in `ast/resolve.rs` perhaps? Ideally the errors in this file are just parse errors rather than semantic errors.

File: crates/wit-parser/src/ast.rs
Comment (alexcrichton) [inline crates/wit-parser/src/ast.rs:139]: The previous split of parsing implicit/explicit items doesn't quite make sense any more with this and I find that the structure here is a bit confusing. Could this instead be modelled as:

1. Parse the `package foo:bar;` at the top of the file, this is unconditionally required to come first.
2. Parse one-item-at-a-time after that which is either a nested package or an ast item.
3. When empty, return

I think that would help simplify a good chunk of the parsing here and in `parse_main_package_items` perhaps? For example an `AstItem` could be expanded to hold `NestedPackage(...)` or something like that. That would AST-wise allow arbitrarily nested packages but the AST validator could reject that saying that nested packages are only supported at the top level.

**Comment by alexcrichton** [comment]
I'll do a final review of this tomorrow and afterwards start to turn the cranks again.

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**
